osquery/requests/requests.h

```diff
@@ -0,0 +1,246 @@
+  std::string destination_;
+
+  /// storage for the serializer reference
+  std::shared_ptr<Serializer> serializer_;
+
+  /// storage for response status
+  Status response_status_;
+
+  /// storage for response parameters
+  boost::property_tree::ptree response_params_;
+};
+
+/**
+ * @brief Abstract base class for serialization implementations
+ *
+ * To define a new serialization mechanism (JSON, XML, etc) for use with
+ * remote connections, subclass osquery::Serializer and implement the pure
+ * virtual methods (as well as any base methods you may want to implement
+ * custom behavior for).
+ */
+class Serializer {
+ public:
+  /**
+   * @brief Set the transport
+   *
+   * @param A transport object
+   */
+  virtual void setTransport(std::shared_ptr<Transport> transport) {
+    transport_ = transport;
+  }
+
+  /**
+   * @brief Returns the HTTP content type, for HTTP/TLS transport
+   *
+   * If a serializer is going to be used for HTTP/TLS, it probably needs to
+   * set it's own content type. Return a string with the appropriate content
+   * type for serializer.
+   *
+   * @return The content type
+   */
+  virtual std::string getContentType() = 0;
+
+  /**
+   * @brief Serialize a property tree into a string
+   *
+   * @param params A property tree of parameters
+   *
+   * @param serialized The string to populate the final serialized params into
+   *
+   * @return An instance of osquery::Status indicating the success or failure
+   * of the operation
+   */
+  virtual Status serialize(const boost::property_tree::ptree& params,
+                           std::string& serialized) = 0;
+
+  /**
+   * @brief Deerialize a property tree into a property tree
+   *
+   * @param params A string of serialized parameters
+   *
+   * @param serialized The property tree to populate the final serialized
+   * params into
+   *
+   * @return An instance of osquery::Status indicating the success or failure
+   * of the operation
+   */
+  virtual Status deserialize(const std::string& serialized,
+                             boost::property_tree::ptree& params) = 0;
+
+  /**
+   * @brief Virtual destructor
+   */
+  virtual ~Serializer() {}
+
+ protected:
+  /// storage for the transport reference
+  std::shared_ptr<Transport> transport_;
+};
+
+/**
+ * @brief Request class for making flexible remote network requests
+ */
+template <class TTransport, class TSerializer>
+class Request {
+ public:
+  /**
+   * @brief This is the constructor that you should use to instantiate Request
+   *
+   * @param destination A string of the remote URI destination
+   */
+  Request(const std::string& destination) : destination_(destination) {
+    transport_ = std::make_shared<TTransport>();
+    transport_->setDestination(destination_);
+    serializer_ = std::make_shared<TSerializer>();
+
+    transport_->setSerializer(serializer_);
+    serializer_->setTransport(transport_);
+  }
+
+  /**
+   * @brief Class destructor
+   */
+  ~Request() {}
```

Comment:
Oh, it seems like a `class ExpertRequest<TLSTransport, JSONSerialization> {}` could be a thing?